### PR TITLE
Implement backtick syntax for defining commands

### PR DESCRIPTION
**Pull request by moxvallix**
Related: #7 

Implements backtick syntax, replacing the forward slash syntax.
With this syntax multi-line commands are possible, and the lexer removes excess whitespace
when not within a string inside the command.

This is the first step towards having a command literal, but right now it still requires a newline first.

### CODE REVIEW

File: TODO.md [hunk 1]
```diff
@@ -2,11 +2,13 @@
 - `@condition` builtin
 - Builtin functions at module level
 - More permissive resources
+  - Resolve `~` and allow `:`?
 - `break` and `continue`
 - Project config
 - Integrate stdlib
 - Add more functionality to static expressions
 - Pipe operator (`|>`) and underscore (`_`) syntax
+- Inline compile time variables (`data modify storage &foo` etc)
```
Comment (GearsDatapacks) [inline TODO.md:11]: This has been implemented now :)

File: src/lexer/mod.rs [hunk 2]
```diff
@@ -106,9 +110,9 @@ impl Lexer {
       if !self.tokenise_json() {
         value = Some(self.src[position + 1..self.position - 1].into());
       }
-    } else if self.current() == '/' && self.is_newline {
+    } else if self.current() == '`' && self.is_newline {
```
Comment (GearsDatapacks) [inline src/lexer/mod.rs:113]: We don't need to check for a newline here any more, as the backticks are used as delimiters instead

File: src/lexer/mod.rs [hunk 3]
```diff
@@ -456,20 +460,30 @@ impl Lexer {
...
 
-  fn parse_command(&mut self) -> Result<Vec<Token>> {
+  fn parse_command(&mut self, backtick: bool) -> Result<Vec<Token>> {
     let mut tokens = Vec::new();
 
     let mut current_part = EcoString::new();
     let mut line = self.line;
     let mut column = self.column;
+    let mut string_char: Option<char> = None;
+    let mut last_was_whitespace: bool = false;
```
Comment (GearsDatapacks) [inline src/lexer/mod.rs:470]: What do `string_char` and `last_was_whitespace` do? It's not very clear to me

File: src/lexer/mod.rs [hunk 4]
```diff
@@ -540,10 +554,33 @@ impl Lexer {
...
+            string_char = Some(self.current());
+          }
           current_part.push(self.consume());
         }
+        (current, _) => {
+          if current == '\n' {
+            current_part.push(' ');
+            self.consume();
+          } else if current.is_ascii_whitespace() {
+            if last_was_whitespace && !string_char.is_some()  {
```
Comment (GearsDatapacks) [inline src/lexer/mod.rs:572]:
```suggestion
            if last_was_whitespace && string_char.is_none()  {
```

File: src/lexer/token.rs [hunk 5]
```diff
@@ -43,7 +43,7 @@ pub enum TokenKind {
   ReturnKeyword,
 
   // Non-zoglin
-  CommandBegin,
+  CommandBegin(bool),
```
Comment (GearsDatapacks) [inline src/lexer/token.rs:46]: Do we need to store this `bool` in the token? It doesn't seem like it's used anywhere

**Review by GearsDatapacks — COMMENTED** [review]
Looks pretty good! I've left couple of notes and questions inline

File: src/lexer/token.rs (hunk 5 above)
Comment (moxvallix) [reply src/lexer/token.rs:46]: It's used in `tokenise` to determine whether the command literal uses a backtick or is just a keyword command.

Comment (moxvallix) [reply src/lexer/token.rs:46]: It probably can and should be refactored, but that was the easiest implementation without changing how commands were set to be lexed.

File: src/lexer/mod.rs (hunk 3 above)
Comment (moxvallix) [reply src/lexer/mod.rs:475]: I have added comments to explain the two fields. Basically `string_char` is for remembering the character used to open a string in a command, and `last_was_whitespace` is used for whitespace stripping.

**Review by GearsDatapacks — APPROVED** [review]
LGTM!

**Pull request merged**